### DISCUSSION TITLE
appwrite behind nginx reverse proxy saves network's default gateway IP

### DISCUSSION THREAD

**Discussion by VictorJosso** [appwrite/appwrite#4587] [Q&A] [answered]
### 👟 Reproduction steps

I run appwrite on a non-dedicated server, so I changed port mapping in docker, because ports 80 and 443 are already in use. 
Then I use Nginx (which is my main http server) to serve appwrite over `appwrite.domain.com`. 
It works, but appwrite saves `172.21.0.1` (which is the network's default gateway) as the IP address of every client. 
It may cause problems because of rate limiting. 

Here is my nginx config file : 

```
server {
       listen         80;
       server_name    appwrite.domain.com;
       return         301 https://$server_name$request_uri; #Redirection
}


server {
    # listen on port 433 (https)
    listen 443 ssl;
    server_name appwrite.domain.com;

    ssl_certificate /etc/letsencrypt/live/appwrite.domain.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/appwrite.domain.com/privkey.pem;

    access_log /var/log/appwrite_access.log;
    error_log /var/log/appwrite_error.log;


    location / {


        proxy_pass http://127.0.0.1:8031;
        proxy_redirect off;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-Port $server_port;
    }
}
```

### 👍 Expected behavior

It should save the client's real IP address

### 👎 Actual Behavior

It saves `172.21.0.1` as every client's IP address (which is the network's default gateway IP)

### 🎲 Appwrite version

Version 0.14.x

### 💻 Operating system

Linux

### 🧱 Your Environment

Self hosted, custom server build. 
Runs on docker, inside Openmediavault (debian)

### 👀 Have you spent some time to check if this issue has been raised before?

- [x] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by stnguyen90** [comment]
The problem is probably traefik not trusting your nginx so it ignores those proxy headers. I would recommend creating a docker-compose.override.yml file and override the traefik commands to add a flag to trust the reverse proxy. See these docs: https://doc.traefik.io/traefik/v2.1/routing/entrypoints/#forwarded-header.

Fyi, I highly suggest forwarding http and https traffic to Appwrite on http and https ports respectively and letting Appwrite handle the https redirect. Cookies may not work properly if you forward all traffic to http Appwrite.

**Comment by roo12312** [comment]
@VictorJosso Hi any updates on this? even im facing the same issue where the ip 172.21.0.1 is getting logged

**Comment by stnguyen90** [comment]
> @VictorJosso Hi any updates on this? even im facing the same issue where the ip 172.21.0.1 is getting logged

Do you have another proxy in front of Appwrite? If so, did you see my [comment](https://github.com/appwrite/appwrite/issues/3334#issuecomment-1148211930)? If not, what's your setup?

**Comment by zedpr0** [comment]
> 

I am also facing the exact same situation, appwrite behind Nginx, the Nginx has the same conf file as @VictorJosso and I have added the following flags to the docker-compose file:      
      - --entryPoints.web.forwardedHeaders.trustedIPs=172.21.0.1/16
      - --entryPoints.websecure.forwardedHeaders.trustedIPs=172.21.0.1/16
      - --entryPoints.web.proxyProtocol.trustedIPs=172.21.0.1/16
      - --entryPoints.websecure.proxyProtocol.trustedIPs=172.21.0.1/16
yet appwrite is still showing 172.21.0.1 as the ip for every client. I am lost. please do help.

**Comment by VictorJosso** [comment]
@rohankm @zedpr0 please try adding this to your traefik configuration : 

```
- --entryPoints.appwrite_web.forwardedHeaders.trustedIPs=172.21.0.1
```

It worked for me

**Comment by zedpr0** [comment, marked as answer]
oh thank you so much @VictorJosso,

just for anyone else's reference if they are using https like me, I also had to add the https version:

      - --entryPoints.appwrite_web.forwardedHeaders.trustedIPs=172.21.0.1
      - --entryPoints.appwrite_websecure.forwardedHeaders.trustedIPs=172.21.0.1

Have a great day!

**Comment by roo12312** [comment]
> @rohankm @zedpr0 please try adding this to your traefik configuration :
> 
> ```
> - --entryPoints.appwrite_web.forwardedHeaders.trustedIPs=172.21.0.1
> ```
> 
> It worked for me

Thanks bother its working

**Comment by eldadfux** [comment]
If anyone reached this issue from search engines, this post can also help: https://dev.to/appwrite/learn-how-to-run-appwrite-with-your-own-custom-proxy-or-load-balancer-28k
